[IR] Initial introduction of memset_pattern

asb opened the pull request:
Supersedes the draft PR #94992, taking a different approach following feedback:
* Lower in PreISelIntrinsicLowering
* Don't require that the number of bytes to set is a compile-time constant
* Define llvm.memset_pattern rather than llvm.memset_pattern.inline

As discussed in the [RFC thread](https://discourse.llvm.org/t/rfc-introducing-an-llvm-memset-pattern-inline-intrinsic/79496), the intent is that the intrinsic will be lowered to loops, a sequence of stores, or libcalls depending on the expected cost and availability of libcalls on the target. Right now, there's just a single lowering path that aims to handle all cases. My intent would be to  follow up with additional PRs that add additional optimisations when possible (e.g. when libcalls are available, when arguments are known to be constant etc).

topperc commented:
Does this need to care about big endian vs little endian? I notice LoopIdiomRecognize.cpp only handles little endian targets.

asb committed:
[IR] Initial introduction of memset_pattern

asb commented:
> Does this need to care about big endian vs little endian? I notice LoopIdiomRecognize.cpp only handles little endian targets.

You're right, I've adjusted the current implementation to bail out if targeting a big endian system.

With additional tests added and some cleanup now done, I think this is ready for review as is.

asb commented:
Ping - review comments greatly appreciated.

dtcxzyw commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+If the ``isvolatile`` parameter is ``true``, the
+``llvm.memset_pattern`` call is a :ref:`volatile operation <volatile>`. The
+detailed access behavior is not very cleanly specified and it is unwise to
+depend on it.
```
I would like to drop this argument and forbid transforming volatile stores into `llvm.memset_pattern` in `LoopIdiomRecognize`.

dtcxzyw replied:
Oh, I see `MemIntrinsic` needs `isvolatile` to be the fourth argument. I don't insist on this.

asb replied:
Ack, I've left it as-is for now.

dtcxzyw commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1249,6 +1252,21 @@ class MemSetInlineInst : public MemSetInst {
+class MemSetPatternInst : public MemSetInst {
+public:
+  ConstantInt *getLength() const {
+    return cast<ConstantInt>(MemSetInst::getLength());
```
IIRC the constant length restriction has been removed.

See also https://discourse.llvm.org/t/rfc-introducing-an-llvm-memset-pattern-inline-intrinsic/79496/5.

asb replied:
Yes this was a holdover from before I'd removed that restriction. Now removed.

dtcxzyw commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -456,6 +456,109 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+  BasicBlock *NewBB = OrigBB->splitBasicBlock(InsertBefore, "split");
+  BasicBlock *LoopBB =
+      BasicBlock::Create(F->getContext(), "storeloop", F, NewBB);
+  BasicBlock *RemCheckBB =
```
Can we avoid remcheck loop when we already know the length is a multiple of `PatternSize`?

asb replied:
Yes, though my intent was to generate the most generic "dumb" expansion that doesn't take advantage of any of these opportunities to simplify, use #98311 as a pipecleaner for that and then incrementally add such improvements. Happy to do it otherwise if you prefer though.

dtcxzyw replied:
I am ok with leaving it as a todo.

asb committed:
Tweak wording in Langref description based on feedback

asb committed:
Removing ConstantInt getLength (holdover from when this was a restriction)

asb committed:
Properly update memset-pattern.ll test cases

asb commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+"""""""
+
+This is an overloaded intrinsic. You can use ``llvm.memset_pattern`` on
+any integer bit width and for different address spaces. Not all targets
```
I've redrafted this, let me know what you think.

asb commented:
Loked at this again now the 19.x branch has taken place, and all comments are now addressed I believe.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
 
+.. _int_memset_pattern:
+
+'``llvm.memset_pattern``' Intrinsic
```
```suggestion
'``llvm.memset.pattern``' Intrinsic
```

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+
+::
+
+      declare void @llvm.memset_pattern.p0.i64.i128(ptr <dest>, i128 <val>,
```
```suggestion
      declare void @llvm.memset.pattern.p0.i64.i128(ptr <dest>, i128 <val>,
```

nikic reviewed (commented):
Something I'm still not completely sold on is that the size is specified in bytes. I think the only advantage of doing that is that it directly maps onto the apple libcalls. On the flip side, it has the disadvantage that we may end up generating an unnecessary tail loop.

However, the larger conceptual problem I see is that it may prevent future generalization (though I kind of wonder whether that should not be supported right away) to non-integer types. memset.pattern is a good opportunity to perform stores with the correct types. E.g. consider filling memory with non-integral pointers -- in that case you can't just perform the memset after a ptrtoint. And a store of only part of a non-integral pointer would be pretty ill-defined. To a lesser degree this problem also exists for integral pointers, due to provenance considerations.

I'd rather not end up having to need another separate intrinsic to satisfy that case.

asb commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+Overview:
+"""""""""
+
+The '``llvm.memset_pattern.*``' intrinsics fill a block of memory with
```
The intrinsic is indeed named `llvm.memset_pattern`, so these do match the implementation in the current PR:

```
def int_memset_pattern
    : Intrinsic<[],
      [llvm_anyptr_ty, llvm_anyint_ty, llvm_anyint_ty, llvm_i1_ty],
      [IntrWriteMem, IntrArgMemOnly, IntrWillReturn, IntrNoFree, IntrNoCallback,
       NoCapture<ArgIndex<0>>, WriteOnly<ArgIndex<0>>,
       ImmArg<ArgIndex<3>>], "llvm.memset_pattern">;
```

Did you want to advocate for instead calling it `llvm.memset.pattern`?

arsenm replied:
Yes. I do not know why we have a handful of intrinsics that go out of their way to use an _ separator in the name, but we shouldn't add any more of them

asb replied:
Ok, I've gone ahead and renamed to `llvm.memset.pattern`

asb committed:
Removed outdated comment

asb committed:
Change to memset_pattern taking a count rather than a number of bytes

Also simplify/remove tests that are no longer relevant given we don't
need to worry about writes of a partial pattern.

asb committed:
Rename to llvm.memset.pattern as requested in review

asb commented:
> Something I'm still not completely sold on is that the size is specified in bytes. I think the only advantage of doing that is that it directly maps onto the apple libcalls. On the flip side, it has the disadvantage that we may end up generating an unnecessary tail loop.
> 
> However, the larger conceptual problem I see is that it may prevent future generalization (though I kind of wonder whether that should not be supported right away) to non-integer types. memset.pattern is a good opportunity to perform stores with the correct types. E.g. consider filling memory with non-integral pointers -- in that case you can't just perform the memset after a ptrtoint. And a store of only part of a non-integral pointer would be pretty ill-defined. To a lesser degree this problem also exists for integral pointers, due to provenance considerations.
> 
> I'd rather not end up having to need another separate intrinsic to satisfy that case.

I think this is a good call - thank you. I've gone ahead and made that change.

One awkwardness I want to flag for review is that MemSetPatternInst still inherits from MemSetInst which inherits from `MemSetBase`. The awkwardness is that `getLength` returns the length in bytes for the other mem* intrinsics, but it's the number of pattern sets for memset.pattern.

arsenm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1003,6 +1003,14 @@ def int_memset_inline
        NoCapture<ArgIndex<0>>, WriteOnly<ArgIndex<0>>,
        ImmArg<ArgIndex<3>>]>;
 
+// Memset variant that writes a given pattern.
```
Comment what the operands are

asb replied:
Added with inline comments (there's not a totally consistent pattern for describing args - many intrinsics have no description at all, but I see some other examples in the file using inline comments for the args)

arsenm replied:
Really we ought to have tablegen mandatory doc strings

arsenm commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1003,6 +1003,14 @@ def int_memset_inline
+def int_memset_pattern
+    : Intrinsic<[],
+      [llvm_anyptr_ty, llvm_anyint_ty, llvm_anyint_ty, llvm_i1_ty],
+      [IntrWriteMem, IntrArgMemOnly, IntrWillReturn, IntrNoFree, IntrNoCallback,
```
DefaultAttrIntrinsic would hide most of these

asb replied:
I dug through this and unfortunately I can't make use of `DefaultAttrIntrinsic` because `nosync` isn't necessarily true. https://reviews.llvm.org/D86021 switched over memset to using DefaultAttrIntrinsic but this was later backed out in a888e492f601644f44b843373039966e7acacd31 due to `nosync` not applying unconditionally.

bjope commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+to some boundary, this can be specified as an attribute on the argument.
+
+If ``<len>`` is not an integer multiple of the pattern width in bytes, then any
+remainder bytes will be copied from ``<val>``.
```
Could be worth describing how this works with respect to endianess.
I imagine that for a big endian target the `<val>` pattern is specified as a big endian number, and bytes are written by picking bytes round robin starting at the most significant byte in `<val>`. While for litte-endian targets the first byte written is the least significant byte in `<val>`.

asb replied:
I've added what aims to be a minimal description. Also fixed the outdated references to byte counts from an earlier iteration of the intrinsic.

bjope commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+Semantics:
+""""""""""
+
+The '``llvm.memset.pattern*``' intrinsic fills "len" bytes of memory
```
Here it says that `len` is bytes of memory, while above it says it is the number of times the pattern should be repeated. Confusing. But maybe we haven't settled if the pattern is allowed to be non-byte-sized or not.

asb replied:
Thanks, that had indeed been missed and has now been addressed.

bjope commented:
FWIW, when it comes to regular memset out downstream target (with 16 bit char) is treating the length specifier as "number of chars" (just like it is for targets with 8 bit bytes).
However, we have needed to make the value argument an overloaded type in the intrinsic, to allow using i16 as type for the value. This of course result in multiple downstream diffs all over the place, which always have been a minor annoyance.
Theoretically, in the long run, it would be nice if "memset" just was a subset of "memset.pattern" whit the pattern size matching the byte size. At a glance it seems like memset.pattern might fill the gap, but then all optimizations should deal with byte sized value argument in memset.pattern just as well as ordinary memset. I doubt that it will happen in the near future (or ever).
'm just saying that the existing "llvm.memset" is one of those things that is complicated to handle for certain targets. And it wouldn't hurt if memset.pattern would work well regardless of "size of char"/"bits per byte".

asb committed:
Add comments to memset_pattern intrinsic to describe args

asb committed:
Improve memset.pattern langref: fix outdated refs to bytes and mention endianness

asb commented:
I believe all review comments are now addressed. Thank you.

@bjope yes I agree it could be useful for that case. I don't think there's anything in the current description that would prevent moving in the direction you suggest in the future.

nikic commented:
> Theoretically, in the long run, it would be nice if "memset" just was a subset of "memset.pattern" whit the pattern size matching the byte size. At a glance it seems like memset.pattern might fill the gap, but then all optimizations should deal with byte sized value argument in memset.pattern just as well as ordinary memset. I doubt that it will happen in the near future (or ever).

An alternative we should probably consider here is to not have a separate memset.pattern at all, and instead have memset be memset.pattern. If we're going to add a type overload to memset anyway for non-8-bit targets, there's probably not a lot of point to having a separate memset.pattern?

asb commented:
> > Theoretically, in the long run, it would be nice if "memset" just was a subset of "memset.pattern" whit the pattern size matching the byte size. At a glance it seems like memset.pattern might fill the gap, but then all optimizations should deal with byte sized value argument in memset.pattern just as well as ordinary memset. I doubt that it will happen in the near future (or ever).
> 
> An alternative we should probably consider here is to not have a separate memset.pattern at all, and instead have memset be memset.pattern. If we're going to add a type overload to memset anyway for non-8-bit targets, there's probably not a lot of point to having a separate memset.pattern?

I agree unifying could make sense in the future and could be done either by increasing the scope of memset.pattern to subsume memset, or by modifying memset so it subsumes memset.pattern. Are you happy to defer this to the future or do you see implementing such a unification as a blocker for this initial introduction? My assumption so far has been that introducing this intrinsic as an incremental step is likely better due to being non-invasive to existing users and frontends, even if we later modify memset to make memset.pattern redundant.

nikic commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -456,6 +456,56 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+  if (DL.isBigEndian())
+    report_fatal_error("memset.pattern expansion not currently "
+                       "implemented for big-endian targets",
+                       false);
```
Why this check? Your implementation is endian-ness independent. The endian-specific handling is inside the store.

asb replied:
Now removed - this was a holdover from an earlier implementation.

nikic commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -456,6 +456,56 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+
+  if (!isPowerOf2_32(SetValue->getType()->getScalarSizeInBits()))
+    report_fatal_error("Pattern width for memset_pattern must be a power of 2",
+                       false);
```
This limitation would have to be checked by the IR verifier. I don't think it's needed though, your code should work fine without it.

asb replied:
I've removed this limitation and added test coverage.

nikic commented on llvm/test/CodeGen/RISCV/memset-pattern.ll:
```diff
@@ -0,0 +1,253 @@
+; RV64-BOTH-NEXT:    bne a0, a3, .LBB0_1
+; RV64-BOTH-NEXT:  # %bb.2: # %split
+; RV64-BOTH-NEXT:    ret
+  tail call void @llvm.memset.pattern.p0.i64.i128(ptr align 8 %a, i128 %value, i64 1, i1 0)
```
```suggestion
  tail call void @llvm.memset.pattern(ptr align 8 %a, i128 %value, i64 1, i1 0)
```

The mangling here is incorrect anyway -- omit it.

nikic commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -456,6 +456,56 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+  LoopCount->addIncoming(Count, OrigBB);
+
+  // Create the store instruction for the pattern
+  LoopBuilder.CreateAlignedStore(SetValue, CurrentDst, DstAlign, IsVolatile);
```
Doesn't this also need the "PartAlign" logic? It's DstAlign on the first iteration, but afterwards its the common alignment of DstAlign and the stride.

asb replied:
Good catch, thanks. Logic fixed and test coverage for expected alignment added.

nikic commented:
> > An alternative we should probably consider here is to not have a separate memset.pattern at all, and instead have memset be memset.pattern. If we're going to add a type overload to memset anyway for non-8-bit targets, there's probably not a lot of point to having a separate memset.pattern?
> 
> I agree unifying could make sense in the future and could be done either by increasing the scope of memset.pattern to subsume memset, or by modifying memset so it subsumes memset.pattern. Are you happy to defer this to the future or do you see implementing such a unification as a blocker for this initial introduction? My assumption so far has been that introducing this intrinsic as an incremental step is likely better due to being non-invasive to existing users and frontends, even if we later modify memset to make memset.pattern redundant.

Not sure. My thinking here is that you are already adding memset_pattern as part of MemSetInst, so from a C++ API perspective, they become essentially the same thing, and we already have to review all the MemSetInst code to make sure it works with the "pattern" semantics. And I do think that adding it to MemSetInst is the right thing to do. I think the memset lowering code also ends up being exactly the same between memset/memset_pattern, as the distinction is all in the store lowering. So overall, it may be simpler to implement this by adding an extra overload to memset, rather than adding a new intrinsic. I don't want to hold this PR up forever either though :)

asb committed:
Fix incorrect mangling in LangRef and explain memory address is incremented by allocation size

asb committed:
Excise errant memset_pattern mention

asb committed:
Allow memset.pattern expansion for big endian targets

The refusal to do so is a holdover from an older expansion strategy and
semantics.

asb committed:
Remove unnecessary comment

asb committed:
Allow non-power-of-two length patterns

asb committed:
Rename memset-pattern-inline.ll test to memset-pattern.ll to reflect current naming

asb committed:
Remove unnecessary and incorrect mangling from llvm.memset.pattern uses

asb committed:
Fix logic for alignment of stores in memset.pattern expansion

asb commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,62 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+Syntax:
+"""""""
+
+This is an overloaded intrinsic. You can use ``llvm.memset_pattern`` on
```
Thanks, now fixed.

asb commented on llvm/docs/LangRef.rst:
```diff
@@ -15230,6 +15230,64 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+on the argument. The pattern fills will respect the endianness of the target:
+i.e. on little endian targets, the least significant byte of the pattern is
+first in memory, while the most significant byte is first in memory for big
+endian targets.
```
I've added text that intends to convey this.

asb commented:
All comments now addressed I believe, with the only known open issue being whether to proceed in the direction of adding this intrinsic or to move straight to trying to overload memset.

preames commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -456,6 +456,48 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
   ElseTerm->eraseFromParent();
 }
 
+static void createMemSetPatternLoop(Instruction *InsertBefore, Value *DstAddr,
```
This code looks identical to createMemSetLoop to my eye, is there a reason this can't be shared?

preames commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1219,14 +1220,16 @@ class MemIntrinsic : public MemIntrinsicBase<MemIntrinsic> {
     switch (I->getIntrinsicID()) {
     case Intrinsic::memset:
     case Intrinsic::memset_inline:
+    case Intrinsic::memset_pattern:
```
I'm very uncomfortable with adding memset_pattern in the current form to MemSetInst.  There's a bunch of code which visits MemSetInst, and without auditting all of them it's really hard to know if this is correct.  Do you *need* this for an initial patch?  Or can you reasonable separate an initial implementation without it, and then return to this merging in a second commit?

asb replied:
I've updated so MemSetPatternInst is no longer in the MemSetInst hierarchy. I do use MemSetBase, but that's not accessed externally in the LLVM codebase right now.

This is a good call, looking more closely I see e.g. some AnyMemSetInst references in the codebase I'd be wanting to audit more closely. Thanks.

preames reviewed (commented):
I think I agree with @nikic that long term, what we want here is not a separate memset.pattern, but a type parameter to memset.  However, I'm leery of this being stuck in review indefinitely.  Perfection is the enemy of the good.

Here's my suggestion on a path forward here:
* We add the new intrinsic, but explicitly mark it experimental.  This ensures we can change/remove it freely without backwards compat concerns during the next couple steps.
* At the initial stage, memset.pattern would *not* be a memset (in particular, no longer in MemSetInst).
* We'd implement initial lowering, and optimization in terms of the intrinsic.  Feel through the design, make sure we haven't gotten anything wrong.
* Once we're comfortable, we generalize memset with a type param.  At this point, this is a semantically interesting change, but little new code required (it all exists for the pattern variant).  Initially, update the semantics, and lowering.
* Then switch over the optimizations to generate the memset with type param form and delete the intrinsic.

If we don't like the idea of the experimental stepping stone, another idea would be to generalize the existing memset.inline with a type param, and then work our way to the full memset via that path.

asb commented:
> I think I agree with @nikic that long term, what we want here is not a separate memset.pattern, but a type parameter to memset. However, I'm leery of this being stuck in review indefinitely. Perfection is the enemy of the good.
> 
> Here's my suggestion on a path forward here:
> <snip>

Philip's plan seems sensible to me - just wanted to check if reviewers who have participated so far are (@nikic? @arsenm?). If so I'll go ahead and update along those lines. Thanks!

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -591,6 +633,16 @@ bool llvm::expandMemMoveAsLoop(MemMoveInst *Memmove,
 void llvm::expandMemSetAsLoop(MemSetInst *Memset) {
+  if (isa<MemSetPatternInst>(Memset)) {
+    return createMemSetPatternLoop(
+        /* InsertBefore */ Memset,
```
```suggestion
        /* InsertBefore= */ Memset,
```

arsenm commented on llvm/test/CodeGen/RISCV/memset-pattern.ll:
```diff
@@ -0,0 +1,253 @@
+; RV64-BOTH-NEXT:    bne a0, a3, .LBB0_1
+; RV64-BOTH-NEXT:  # %bb.2: # %split
+; RV64-BOTH-NEXT:    ret
+  tail call void @llvm.memset.pattern(ptr align 8 %a, i128 %value, i64 1, i1 0)
```
Missing type mangling part of name

nikic replied:
The mangling suffix is optional in textual IR. We prefer not to add it.

nikic replied:
It does not. The IR parser will automatically determine the correct suffix for each call.

nikic replied:
For reference, I implemented support for that in https://github.com/llvm/llvm-project/pull/89172.

nikic commented:
> I think I agree with @nikic that long term, what we want here is not a separate memset.pattern, but a type parameter to memset. However, I'm leery of this being stuck in review indefinitely. Perfection is the enemy of the good.
> 
> Here's my suggestion on a path forward here:
> 
>   * We add the new intrinsic, but explicitly mark it experimental.  This ensures we can change/remove it freely without backwards compat concerns during the next couple steps.
> 
>   * At the initial stage, memset.pattern would _not_ be a memset (in particular, no longer in MemSetInst).
> 
>   * We'd implement initial lowering, and optimization in terms of the intrinsic.  Feel through the design, make sure we haven't gotten anything wrong.
> 
>   * Once we're comfortable, we generalize memset with a type param.  At this point, this is a semantically interesting change, but little new code required (it all exists for the pattern variant).  Initially, update the semantics, and lowering.
> 
>   * Then switch over the optimizations to generate the memset with type param form and delete the intrinsic.
> 
> If we don't like the idea of the experimental stepping stone, another idea would be to generalize the existing memset.inline with a type param, and then work our way to the full memset via that path.

I'm okay with doing it this way, though I suspect it will increase the effort needed to reach the end state (while making it easier to get an MVP). I think with this direction, we should try to do things in a way that reduce mentioning the intrinsic name. For example have some common base class for memset and memset.pattern and then migrate code using MemSetInst to GenericMemSetInst (placeholder name).

asb committed:
Merge remote-tracking branch 'origin/main' into 2024q2-memset-pattern-inline-preiselintrinsiclowering

asb committed:
Regenerate memset-pattern.ll after merge

asb committed:
Use normal createMemsetAsLoop helper for memset.pattern

As pointed out in the review, with various changes to memset.pattern
semantics since the first version these are now effectively identical.

asb committed:
Move MemSetPattern out of the MemSet hierarchy

asb committed:
Rename to llvm.experimental.memset.pattern

asb committed:
Fix underline length in langref

preames reviewed (approved):
LGTM

topperc commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1263,6 +1263,46 @@ class MemSetInlineInst : public MemSetInst {
+
+  // Methods for support of type inquiry through isa, cast, and dyn_cast:
+  static bool classof(const IntrinsicInst *I) {
+    switch (I->getIntrinsicID()) {
```
`return I->getIntrinsicID() == Intrinsic::experimental_memset_pattern`?

topperc commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1263,6 +1263,46 @@ class MemSetInlineInst : public MemSetInst {
   }
 };
 
+/// This is the base class for memset.pattern
```
experimental.memsest.pattern?

topperc commented on llvm/docs/LangRef.rst:
```diff
@@ -15434,6 +15434,66 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+
+This is an overloaded intrinsic. You can use
+``llvm.experimental.memset.pattern`` on any integer bit width that is an
+integral number of bytes and for different address spaces. Not all targets
```
Do we need a Verifier check to check that the size is integral number of bytes?

asb replied:
I've now added this check to the verifier.

asb replied:
Actually that was my bad. There was some discussion about this before and in the end non-integral bitwidths should work and are tested in `llvm/test/Transforms/PreISelIntrinsicLowering/RISCV/memset-pattern.ll`. I've now pushed a fix to LangRef to remove the stated restriction.

asb committed:
Address review comments

asb committed:
Verkfy llvm.experimental.memset.pattern pattern arg is integral number of bytes

asb committed:
Revert "Verkfy llvm.experimental.memset.pattern pattern arg is integral number of bytes"

This reverts commit 78bad3b075aaf26dd909a84eb45a7bf87a4d4c8f.

asb committed:
Remove outdated comment about integral bit widths only

asb commented:
I think I'm ok to merge with Philip's LGTM and I believe all outstanding comments have been addressed, but I'll give this another day or so as it's been open so long. Please speak up if there's anything you feel wasn't fully reflected in the current code.

s-barannikov commented on llvm/docs/LangRef.rst:
```diff
@@ -15434,6 +15434,65 @@ The behavior of '``llvm.memset.inline.*``' is equivalent to the behavior of
+the target: i.e. on little endian targets, the least significant byte of the
+pattern is first in memory, while the most significant byte is first in memory
+for big endian targets. The memory address is incremented by the allocation
+size of the type.
```
If the size is not an integral number of bytes, what happens with the padding bits?
Do they get unspecified / undefined values or do they preserve their values?
Should this be clarified in LangRef?

nikic replied:
The behavior is the same as whatever stores of the type would do.

Possibly we should just explicit refer to store semantics here and reduce the description to something like this:

> The '`llvm.experimental.memset.pattern*`' intrinsic stores the provided value `<count>>` times, incrementing by the allocation size of the type each time. The stores follows the usual semantics of `store` instructions, including regarding endianness and padding.

asb replied:
Thanks, that's much clearer - I've adopted that wording.

nikic commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1263,6 +1263,41 @@ class MemSetInlineInst : public MemSetInst {
+  static bool classof(const Value *V) {
+    return isa<IntrinsicInst>(V) && classof(cast<IntrinsicInst>(V));
+  }
+};
```
Why do we need both MemSetPatternIntrinsic and MemSetPatternInst?

asb replied:
I'm mirroring the pattern followed by other mem intrinsics, and although it's not super pretty, having both classes like this (as for the standard MemSet intrinsics) seems the way that reduces copy and paste of accessor code.

nikic replied:
Hm okay, I thought this might only be needed for cases where we have both atomic and non-atomic variants.

nikic reviewed (approved):
LGTM

asb committed:
Adopt Nikita's langref rewording suggestion

dtcxzyw commented on llvm/include/llvm/IR/IntrinsicInst.h:
```diff
@@ -1263,6 +1263,41 @@ class MemSetInlineInst : public MemSetInst {
   }
 };
 
+/// This is the base class for llm.experimental.memset.pattern
```
```suggestion
/// This is the base class for llvm.experimental.memset.pattern
```

asb committed:
typo fixes
